Firestore/core/src/firebase/firestore/api/query_core.mm

```diff
@@ -0,0 +1,327 @@
+using firebase::firestore::core::QueryListener;
+using firebase::firestore::core::ViewSnapshot;
+using firebase::firestore::model::DocumentKey;
+using firebase::firestore::model::FieldPath;
+using firebase::firestore::model::FieldValue;
+using firebase::firestore::model::ResourcePath;
+using firebase::firestore::util::Status;
+using firebase::firestore::util::StatusOr;
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+bool operator==(const Query& lhs, const Query& rhs) {
+  return lhs.firestore() == rhs.firestore() &&
+         objc::Equals(lhs.query(), rhs.query());
+}
+
+size_t Query::Hash() const {
+  return util::Hash(firestore_.get(), [query() hash]);
+}
+
+FSTQuery* Query::query() const {
+  return query_;
+}
+
+void Query::GetDocuments(Source source, QuerySnapshot::Listener&& callback) {
+  if (source == Source::Cache) {
+    [firestore_->client() getDocumentsFromLocalCache:*this
+                                            callback:std::move(callback)];
+    return;
+  }
+
+  ListenOptions options(
+      /*include_query_metadata_changes=*/true,
+      /*include_document_metadata_changes=*/true,
+      /*wait_for_sync_when_online=*/true);
+
+  class ListenOnce : public EventListener<QuerySnapshot> {
+   public:
+    ListenOnce(Source source, QuerySnapshot::Listener&& listener)
+        : source_(source), listener_(std::move(listener)) {
+    }
+
+    void OnEvent(StatusOr<QuerySnapshot> maybe_snapshot) override {
+      if (!maybe_snapshot.ok()) {
+        listener_->OnEvent(std::move(maybe_snapshot));
+        return;
+      }
+
+      QuerySnapshot snapshot = std::move(maybe_snapshot).ValueOrDie();
+
+      // Remove query first before passing event to user to avoid user actions
+      // affecting the now stale query.
+      ListenerRegistration registration =
+          registration_promise_.get_future().get();
+      registration.Remove();
+
+      if (snapshot.metadata().from_cache() && source_ == Source::Server) {
+        listener_->OnEvent(Status{
+            FirestoreErrorCode::Unavailable,
+            "Failed to get documents from server. (However, these documents "
+            "may exist in the local cache. Run again without setting source to "
+            "FirestoreSourceServer to retrieve the cached documents.)"});
+      } else {
+        listener_->OnEvent(std::move(snapshot));
+      }
+    };
+
+    void Resolve(ListenerRegistration&& registration) {
+      registration_promise_.set_value(std::move(registration));
+    }
+
+   private:
+    Source source_;
+    QuerySnapshot::Listener listener_;
+
+    std::promise<ListenerRegistration> registration_promise_;
+  };
+
+  auto listener = absl::make_unique<ListenOnce>(source, std::move(callback));
+  auto listener_unowned = listener.get();
+
+  ListenerRegistration registration =
+      AddSnapshotListener(std::move(options), std::move(listener));
+
+  listener_unowned->Resolve(std::move(registration));
```

Comment:
Not for this PR: this general pattern has a lot of resemblance to `api::DocumentReference::GetDocument`. Perhaps there is a way to abstract some part of it away.

Reply:
Agreed in principle, but things are just different enough that this is not easy.